### ISSUE TITLE
React v18 + Formik + controlled Downshift Select bug

### ISSUE THREAD

**Issue by stevenocchipinti** [downshift-js/downshift#1511]
- `downshift` version: `v7.6.0`
- `node` version: `v18`
- `npm` (or `yarn`) version: `v9.5.0`

**Relevant code or config**

```javascript
useSelect({ selectedItem: ...})
```

**What you did**:
Made a demo with React v18 + Formik + Downshift using a controlled Select component (via the `useSelect` hook).

**What happened**:
An infinite loop with the error message:

> index.js:27 Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.

**Reproduction repository**:
https://codesandbox.io/s/react-v8-formik-downshift-63l7sw?file=/src/App.js

**Problem description**:
I've described the problem and steps to reproduce it in the [codesandbox above](https://codesandbox.io/s/react-v8-formik-downshift-63l7sw?file=/src/App.js).

In short, the problem seems to be limited to:

* React v18 (possibly related to #1384)
* Formik
* A controlled Select
* Using touch events (tested with mobile emulation in chrome dev tools)

**Suggested solution**:
Not sure.

**Comment by oddvernes** [comment]
We are experiencing the same error message in our [autocomplete component](https://storybook.eds.equinor.com/?path=/docs/inputs-autocomplete--docs). It is triggered every time by clicking an item in the dropdown using touch/mobile emulation in chrome/firefox. It can also be triggered when using react-hook-forms with regular mouse/kb after selecting several items. 

- React v18
- Downshift v7.6.0
- both controlled and uncontrolled
- `floating-ui` is used for positioning

the first item in the stacktrace is downshift `function handleRefs`
![bilde](https://github.com/downshift-js/downshift/assets/4499531/a87caee7-7b8d-4924-90d5-42136b0874d7)

**Comment by silviuaavram** [comment]
Thank you for reporting this! I took a look and the sandbox but I did not find any obvious issue with it. I also checked the controlled example in the docsite and it works with touch: https://www.downshift-js.com/use-select#controlling-state

Something must be wrong with using that selected item via formik, I guess. It needs a little more digging, I can do that but it takes more time, so if you want you can also check, since it's a great repro sandbox you have! I am happy to review any bugfix submitted, if any.

I used the stateReducer to try and debug. Apparently in touch mode it triggers itemClick indefinitely. Let me know if you found the issue, I will also post updates but, as I said, they can be delayed. Thank you!

```js
    stateReducer(state, { changes, type }) {
      console.log(state, changes, type);
      return changes;
    }
```

**Comment by daves28** [comment]
Also having this issue but with react-hook-form, I think its not specific to Formik. Similarly only occurring on mobile/touch interfaces.

**Comment by minkiele** [comment]
I'm experiencing the same issue with **React 18.2.0**, **Downshift 7.6.0** and **Formik 2.4.2**. I've been able to work around this problem by wrapping Formik's `setValue` callback in a `setTimeout`. I know that's just a hacky workaround and not a solution but maybe it will give someone some clues on what's really happening.

Using `useState` BTW does not cause any problem.

**Comment by FdelMazo** [comment]
> I've been able to work around this problem by wrapping Formik's setValue callback in a setTimeout

@minkiele I didn't follow you. Could you provide some code for this workaround? ty!

**Comment by minkiele** [comment]
> > I've been able to work around this problem by wrapping Formik's setValue callback in a setTimeout
> 
> @minkiele I didn't follow you. Could you provide some code for this workaround? ty!

This is a minimal, stripped down version of what I did to work around the problem
```js
// Get the value setter from Formik's useField hook
const [,,{setValue}] = useField(fieldName);
// Setup Downshift useSelect Hook
useSelect({
  // Lot of omitted options...
  onSelectedItemChange(({selectedItem}) => {
    // Normally you would immediately run setValue directly in the callback
    setTimeout(() => {
      setValue(selectedItem.value);
    });
  })
})
```
In this example `selectedItem` is an object made this way `{ label: string, value: string}`.

BTW let me restate that *this is not a solution, it's just a hacky workaround* that worked on my project.

**Comment by silviuaavram** [comment]
I am going to try to do an update to how we handle state changes. It may or may not work, it's why I will publish an alpha version with it, so everyone can try it out.

Will update soon. Thanks!

**Comment by silviuaavram** [comment]
Everyone, try `downshift@8.2.4-alpha.0`. If this solves the issue and does not break anything, we will release it as normal 8.2.4. Thank you!

**Comment by minkiele** [comment]
Sorry for answering so late, I tried to swap in this version but there is some API incompatibility between this version and the one I'm using (still v7). As soon as I have time to fix it I'll let you know.

**Comment by silviuaavram** [comment]
We can try looking further into it. It's a different behaviour that was introduced in React 18 apparently, still did not get any resolution by the React team.

**Comment by seedy** [comment]
@silviuaavram can you explain why you just closed the issue ?

**Comment by seedy** [comment]
For all still bumping into the issue :
Resolution from https://github.com/downshift-js/downshift/pull/1558 : upgrade formik to 2.4.2
